### PR TITLE
Update code for chapter 12 NYC trip usecase

### PR DESCRIPTION
**Pull request by krisgeus**
TODO:
- [x] Use compose.override as in other chapters
- [x] Fix data loading for taxi DB
- [x] Fix citbike data loading
- [x] Refactor pandas code
- [x] Figure out why the first two dagruns always fail with the taxi data download (maybe the reboot cron job does not work)

### CODE REVIEW

File: chapter12/dags/nyc_dag.py [hunk 1]
```diff
@@ -104,10 +104,10 @@ def transform_citi_bike_data(df):
             crs="EPSG:4326",
             geometry=geopandas.points_from_xy(df["end_station_longitude"], df["end_station_latitude"]),
         )
-        df_with_zones = geopandas.sjoin(start_gdf, taxi_zones, how="left", op="within").rename(
+        df_with_zones = geopandas.sjoin(start_gdf, taxi_zones, how="left").rename(
             columns={"LocationID": "start_location_id"}
         )
-        end_zones = geopandas.sjoin(end_gdf, taxi_zones, how="left", op="within")
+        end_zones = geopandas.sjoin(end_gdf, taxi_zones, how="left")
         df_with_zones["end_location_id"] = end_zones["LocationID"]
```
Comment (ismaelcv) [inline chapter12/dags/nyc_dag.py:111]: In general the use of Pandas is a bit updated throughout this chapter.
They are all valid and will run,  but could be way better.

For example direct assignments like `df_with_zones["end_location_id"] = end_zones["LocationID"]` should be done with the .assign() WOW. 

For example, the function `transform_taxi_data` could be greatly improved with a good refactoring.

I can definitively help out, lets discuss it about it today in the meeting

File: chapter12/services/taxi_db/convert_pq_to_csv.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+import pyarrow.csv as csv
+import pyarrow.parquet as pq
+
+columns=["tpep_pickup_datetime","tpep_dropoff_datetime","trip_distance","PULocationID","DOLocationID"]
+
+
+DATA_YEAR=os.environ.get("DATA_YEAR", 2023)
+
+for i in range(1, 13):
```
Comment (ismaelcv) [inline chapter12/services/taxi_db/convert_pq_to_csv.py:12]: `for month in range(1,13)`:...

File: chapter12/setup.py [hunk 3]
```diff
@@ -6,6 +6,6 @@
     description="Package containing custom Airflow components for NYC transportation project.",
     packages=setuptools.find_packages(where="src"),
     package_dir={"": "src"},
-    install_requires=["apache-airflow~=2.0.0b3"],
-    python_requires=">=3.7.*",
+    install_requires=["apache-airflow~=2.9"],
+    python_requires=">=3.7",
```
Comment (ismaelcv) [inline chapter12/setup.py:10]: Maybe good to go to 3.12 as the Other dockerfiles?

File: chapter12/dags/nyc_dag.py [hunk 4]
```diff
@@ -14,212 +14,198 @@
...
+        response = requests.get(url, auth=HTTPBasicAuth(citibike_conn.login, citibike_conn.password))
+        data = response.json()
 
-    url = f"http://{citibike_conn.host}:{citibike_conn.port}/recent/minute/15"
-    response = requests.get(url, auth=HTTPBasicAuth(citibike_conn.login, citibike_conn.password))
-    data = response.json()
+        s3_hook = S3Hook(aws_conn_id="s3")
+        s3_hook.load_string(
+            string_data=json.dumps(data),
+            key=f"raw/citibike/{ts_nodash}.json",
```
Comment (ismaelcv) [inline chapter12/dags/nyc_dag.py:34]: all  instances with {{ ts_nodash }}
That should be changed to {{data_interval_start | ts_nodash}}

**Review by danielvdende — APPROVED** [review]

**Pull request merged**